verification/refinement.md

```diff
@@ -1,2 +1,156 @@
+
+  1. A Verilog variable name as a string. If the Verilog variable is in the top module of Verilog, you can omit the module name (it does not hurt to add it). Otherwise, you must specify the complete hierarchy. For example, if you want to refer to a signal `S` in module `A` that is instantiated with name `IA` in the top module, while the top module's instance name is `VERILOG` (specified in the previous section), then you should use `VERILOG.IA.S` to refer to it.
+  2. A predicate that has at least a `=` in it (you can use `>=` , `<=` , `==` and etc.) This is just for our parser to distinguish it from the first case
+  3. A string-string pair that is in the form of a list or map. If it is given as a list, the first element is regarded as the condition and the second element is regarded as the mapping. It conveys the meaning of "when the condition is true, the ILA and Verilog variables should have the mapping". If the condition is not true, there is no mapping guaranteed. If the string-string pair is given as a map, it must have two elements. One element should have the key `cond` and the other should have the  key `map`. The `map` part could be a string of Verilog variable name as in case (1) or a Verilog expression as case (2).
+  4. A list of string-string pairs, each pair follows requirement in (3).
+
+Below are some of the examples:
+```json
+{
+  "state mapping": {
+    "ILA_state_1" : "Verilog_state_1",                                  // case 1
+    "ILA_state_2" : "ILA_state_2 == Verilog_state_2 + Verilog_state_3", // case 2
+    "ILA_state_3" : ["__START__", "Verilog_state_3"],                   // case 3.a
+    "ILA_state_4" : { "cond":"__START__", "map":"Verilog_state_4"},     // case 3.b
+    "ILA_state_5" : [["__START__", "Verilog_state_5"], ["__IEND__","Verilog_state_6"]],     // case 4
+    // more ...
+    }, 
+    // other fields ...
+}
+```
+
+One note in the above example: the condition can refer to special signals (`__START__ ` and `__IEND__`), which are the condition when the instruction-under-verification starts to execute and finishes.
+
+### Instruction Completion Condition
+
+The instruction completion condition is specified per instruction. In the JSON file, it is a list of maps. The list does not need to be a full list of instructions. Those not in the list will not be verified. Each dictionary must have an element whose key is `instruction` and the value of this element is the name of the instruction in the ILA model. Besides this name element, it must contain one of the `ready bound` or the `ready signal` element. The `ready bound`  specifies a bound that the instruction takes. It is used for instructions that take a fixed number of cycles. Alternatively, one can provide a signal (or a predicate) in the `ready signal` field.
+
+There are several optional fields. They are `start condition`, `max bound`, `flush constraint`, `pre flush-end` and `post flush-end`. 
```

Comment:
Is it `pre flush-end` or `pre-flush end`? As per [this line](https://github.com/Bo-Yuan-Huang/ILAng/blob/a7f693660cf3d0335e8dc0049d3ec40f76095255/src/vtarget-out/single_target.cc#L120), it seems to be later...

Reply:
Will modify later.